File: libs/east-diagnostics/src/rules/no-inline-credentials.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import type * as ts from "typescript";
import type { EastRule, TsModule } from "../types.js";
import { importsEastPackage } from "../east-source.js";

const NAME = "no-inline-credentials";
const CODE = 990032;

/** Property names that carry secrets in the io config types (SQL.*, Storage.S3,
 * Transfer.SFTP/FTP, NoSQL.*) plus the generic token-ish names. Name-based on
 * purpose: configs are often built as standalone object literals before any
 * East type is in contextual-type reach, so type-directed matching would miss
 * the common case. `accessKeyId` is included because it always travels with its
 * secret pair and has no legitimate literal form either. */
const CREDENTIAL_FIELDS = new Set([
  "password",
  "passphrase",
  "privateKey",
  "secretAccessKey",
  "accessKeyId",
  "apiKey",
  "token",
  "accessToken",
  "refreshToken",
  "sessionToken",
  "secretKey",
  "clientSecret",
]);

/** Sibling fields that identify the target host — used for the local-endpoint
 * exemption below. */
const HOST_FIELDS = new Set(["host", "hostname", "endpoint", "url", "server", "address"]);

const LOCAL_HOSTS = ["localhost", "127.0.0.1", "::1"];

function isStringy(node: ts.Node, t: TsModule): node is ts.StringLiteral | ts.NoSubstitutionTemplateLiteral {
  return t.isStringLiteral(node) || t.isNoSubstitutionTemplateLiteral(node);
}

/** The literal credential inside `init`, if there is one: a direct non-empty
 * string literal, or one wrapped in an option/variant constructor —
 * `some("…")`, `East.some("…")`, `variant("some", "…")`, `East.variant(…)`. */
function literalCredential(init: ts.Expression, t: TsModule): ts.Node | undefined {
  if (isStringy(init, t) && init.text.length > 0) return init;
  if (!t.isCallExpression(init)) return undefined;
  const callee = init.expression;
  const calleeName = t.isIdentifier(callee)
    ? callee.text
    : t.isPropertyAccessExpression(callee)
      ? callee.name.text
      : undefined;
  if (calleeName !== "some" && calleeName !== "variant") return undefined;
  // For `variant("some", value)` the first argument is the case tag, not data.
  const args = calleeName === "variant" ? init.arguments.slice(1) : init.arguments;
  for (const arg of args) {
    if (isStringy(arg, t) && arg.text.length > 0) return arg;
  }
  return undefined;
}

/** Does any string literal anywhere under `node` name a local host? */
function mentionsLocalHost(node: ts.Node, t: TsModule): boolean {
  if (isStringy(node, t) && LOCAL_HOSTS.some((h) => node.text.includes(h))) return true;
  let found = false;
  t.forEachChild(node, (child) => {
    if (!found && mentionsLocalHost(child, t)) found = true;
  });
  return found;
}

/** Is a sibling host/endpoint field of this object literal pointing at a local
 * address? Credentials for a developer's own localhost container (test
 * databases, MinIO, …) are not secrets — flagging them would only teach people
 * to ignore the rule. */
function targetsLocalHost(obj: ts.ObjectLiteralExpression, t: TsModule): boolean {
  for (const prop of obj.properties) {
    if (!t.isPropertyAssignment(prop)) continue;
    const name = t.isIdentifier(prop.name) || t.isStringLiteral(prop.name) ? prop.name.text : undefined;
    if (name !== undefined && HOST_FIELDS.has(name) && mentionsLocalHost(prop.initializer, t)) return true;
  }
  return false;
}

// East/e3 source is compiled to IR that is content-addressed, stored in
// package objects, exported, cached, and replicated — a literal credential in
// it is effectively unredactable, and it is readable by anyone with repo
// access. The io config types force the issue (SQL passwords are REQUIRED
// string fields), so the correct form is name-in-IR / value-at-runtime:
// `Env.get("NAME")` (east-node-std / east-py-std), with the value supplied by
// the environment — the developer's shell locally, the deployment's secret
// store in hosted runtimes.
export const noInlineCredentials: EastRule = {
  name: NAME,
  code: CODE,
  description:
    "Flag literal string credentials (password, secretAccessKey, token, …) in East/e3 source — use Env.get so the IR carries the variable name, not the secret.",
  check(node, ctx) {
    const t = ctx.ts;
    if (!t.isPropertyAssignment(node)) return;
    const name = t.isIdentifier(node.name) || t.isStringLiteral(node.name) ? node.name.text : undefined;
    if (name === undefined || !CREDENTIAL_FIELDS.has(name)) return;
    if (!importsEastPackage(ctx.sourceFile, t)) return;

    const literal = literalCredential(node.initializer, t);
    if (literal === undefined) return;
    if (t.isObjectLiteralExpression(node.parent) && targetsLocalHost(node.parent, t)) return;

    const sf = ctx.sourceFile;
    const start = literal.getStart(sf);
    ctx.report({
      ruleName: NAME,
      code: CODE,
      start,
      length: literal.getEnd() - start,
      messageText:
        `Literal credential in \`${name}\` — East source compiles to content-addressed IR that is stored, exported, and replicated, so a secret here is effectively unredactable. Use \`Env.get("YOUR_VAR")\` (east-node-std) and supply the value per environment: your shell locally, the deployment's secret store in hosted runtimes.`,
      category: "warning",
    });
  },
};

File: libs/east-py/packages/east-py/east/diagnostics/rules/no_inline_credentials.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""``no-inline-credentials``: East source compiles to IR that is
content-addressed, stored, exported, cached and replicated — a literal secret
in it is effectively unredactable, and readable by anyone with repo access.
The io config types force the issue (a SQL password is a required field), so
the shape that works is name-in-IR, value-at-runtime: ``east_py_std``'s
``env_get("PGPASSWORD")`` in a body, or ``os.environ`` inside a platform
function — supplied by the shell locally and by the deployment's secret store
in a hosted runtime.

Credentials pointed at a local endpoint are exempt: a developer's own
localhost container is not a secret, and flagging it only teaches people to
ignore the rule. So is a value that is plainly a NAME rather than a secret —
``{"api_key": "X-Api-Key"}`` is a header table, ``"${TOKEN}"`` a template —
and a script's ``if __name__ == "__main__":`` block. The TypeScript rule of
the same name.
"""

from __future__ import annotations

import ast
import re

from east.diagnostics.scope import main_guard_node_ids
from east.diagnostics.types import Body, Context, body_nodes

#: the config fields that carry a secret, by name — a config is usually built
#: as a plain dict, far from any declared East type, so the name is the tell
CREDENTIAL_FIELDS = frozenset({
    "password", "passphrase", "private_key", "privateKey", "secret_access_key", "secretAccessKey",
    "access_key_id", "accessKeyId", "api_key", "apiKey", "token", "access_token", "accessToken",
    "refresh_token", "refreshToken", "session_token", "sessionToken", "secret_key", "secretKey",
    "client_secret", "clientSecret",
})
#: sibling fields naming the target, for the local-endpoint exemption
HOST_FIELDS = frozenset({"host", "hostname", "endpoint", "url", "server", "address"})
LOCAL_HOSTS = ("localhost", "127.0.0.1", "::1")
#: a header / field NAME (`X-Api-Key`, `Authorization`) or a substitution
#: template (`${TOKEN}`, `{{ token }}`, `%(token)s`) — not a secret
_NAME_LIKE = re.compile(r"[A-Z][A-Za-z]*(?:-[A-Z][A-Za-z]*)*")
_TEMPLATE = re.compile(r"\$\{.*\}|\{\{.*\}\}|%\(.*\)s")


def message(field: str) -> str:
    return (f"literal credential in `{field}` — East source compiles to content-addressed IR that "
            "is stored, exported and replicated, so a secret here cannot be redacted. Read it at "
            'runtime instead: east_py_std\'s env_get("YOUR_VAR") in a body, or os.environ inside a '
            "platform function")


class NoInlineCredentials:
    name = "no-inline-credentials"
    code = 20
    category = "warning"
    description = ("No literal credentials in East source — read them at runtime so the IR carries "
                   "the variable name, not the secret.")

    def check(self, body: Body, ctx: Context) -> None:
        # `body_nodes`, not `ast.walk(body.node)`: a nested body is visited as a
        # body of its own, so walking into it would report every finding twice.
        for node in body_nodes(body):
            self._dict(node, ctx)

    def check_module(self, ctx: Context) -> None:
        script = main_guard_node_ids(ctx)
        for node in ast.walk(ctx.tree):
            if id(node) not in script and not ctx.in_body(node):
                self._dict(node, ctx)

    def _dict(self, node: ast.AST, ctx: Context) -> None:
        if not isinstance(node, ast.Dict) or _targets_local_host(node):
            return
        for key, value in zip(node.keys, node.values, strict=False):
            if not (isinstance(key, ast.Constant) and isinstance(key.value, str)):
                continue
            if key.value not in CREDENTIAL_FIELDS:
                continue
            literal = _literal_secret(value)
            if literal is not None:
                ctx.report(literal, self, message(key.value))


def _literal_secret(node: ast.AST) -> ast.AST | None:
    """The non-empty string literal inside ``node`` that reads as a secret,
    directly or wrapped in ``some(...)`` / ``variant("some", …)``."""
    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        return node if _is_secret_text(node.value) else None
    if not isinstance(node, ast.Call):
        return None
    func = node.func
    name = func.attr if isinstance(func, ast.Attribute) else getattr(func, "id", "")
    if name not in ("some", "variant"):
        return None
    args = node.args[1:] if name == "variant" else node.args
    for arg in args:
        if isinstance(arg, ast.Constant) and isinstance(arg.value, str) and _is_secret_text(arg.value):
            return arg
    return None


def _is_secret_text(value: str) -> bool:
    if not value:
        return False
    return _NAME_LIKE.fullmatch(value) is None and _TEMPLATE.fullmatch(value) is None


def _targets_local_host(node: ast.Dict) -> bool:
    """Whether a sibling host field of this dict names a local address."""
    for key, value in zip(node.keys, node.values, strict=False):
        if not (isinstance(key, ast.Constant) and key.value in HOST_FIELDS):
            continue
        for inner in ast.walk(value):
            if isinstance(inner, ast.Constant) and isinstance(inner.value, str) \
                    and any(h in inner.value for h in LOCAL_HOSTS):
                return True
    return False
